docs/architecture/grpc-for-wcf-developers/migrating-duplex-services.md

```diff
@@ -1,11 +1,481 @@
+
+The stream is passed to an async `DisplayAsync` method; the application then waits for the user to press a key, then cancels the `DisplayAsync` method and waits for the task to complete before exiting.
+
+> [!NOTE]
+> This code is using the new C# 8 "using declaration" syntax here to dispose of the stream and the channel when the `Main` method exits. It's a small change, but a nice one that reduces indentations and empty lines.
+
+#### Consuming the stream
+
+WCF used callback interfaces to allow the server to call methods directly on the client. gRPC streams work differently; the client iterates over the returned stream and processes messages, just as though they were returned from a local method returning an `IEnumerable`.
+
+The `IAsyncStreamReader<T>` type works much like an `IEnumerator<T>`: there is a `MoveNext` method that will return true as long as there is more data, and a `Current` property that returns the latest value. The only difference is that the `MoveNext` method returns a `Task<bool>` instead of just a `bool`. The `ReadAllAsync` extension method wraps the stream in a standard C# 8 `IAsyncEnumerable` that can be used with the new `await foreach` syntax.
+
+```csharp
+static async Task DisplayAsync(IAsyncStreamReader<StockTickerUpdate> stream, CancellationToken token)
+{
+    try
+    {
+        await foreach (var update in stream.ReadAllAsync(token))
+        {
+            Console.WriteLine($"{update.Symbol}: {update.Price}");
+        }
+    }
+    catch (RpcException e) when (e.StatusCode == StatusCode.Cancelled)
+    {
+        return;
+    }
+    catch (OperationCanceledException)
+    {
+        Console.WriteLine("Finished.");
+    }
+}
+```
+
+> [!TIP]
+> In the section at the end of the chapter, there will be a look at how to add an extension method and classes to wrap `IAsyncStreamReader<T>` in an `IObservable<T>` for developers using reactive programming patterns.
+
+Again, be careful to catch exceptions here because of the possibility of network failure (as well as the `OperationCanceledException` that will inevitably be thrown because the code is using a `CancellationToken` to break the loop). The `RpcException` type has a lot of useful information about gRPC runtime errors, including the `StatusCode` (see [Chapter X section Y]()).
+
+## Bi-directional Streaming
+
+A WCF Full Duplex service allows for asynchronous, real-time messaging in both directions. In the server streaming example, the client simply initiates a request, then receives a stream of updates. A better version of that service would allow the client to add and remove stocks from the list without needing to stop and create a new subscription. That functionality has been implemented in the [FullStockTickerServer](https://github.com/RendleLabs/grpc-for-wcf-developers/tree/master/FullStockTickerSample/wcf/FullStockTicker) sample.
```

Comment:
should this be moved to dotnet/samples instead?

Reply:
Yes, probably.